{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import sqlite3\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib as mpl\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "\n",
    "\n",
    "def idxwhere(x):\n",
    "    return x[x].index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "color_map = {'acarbose': 'goldenrod', 'control': 'darkblue',\n",
    "             'UM': 'darkblue', 'UT': 'darkgreen',\n",
    "             'male': 'blue', 'female': 'magenta',\n",
    "             'C2013': 'blue', 'Glenn': 'red',\n",
    "             'B1A': 'lightblue', 'B1B': 'lightgreen',\n",
    "             'untrusted': 'grey',\n",
    "             0: 'white', 1: 'black'}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "con = sqlite3.connect('data/core.muri2.2.denorm.db')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cvrg_ratio_B1A = pd.read_sql(\n",
    "    \"\"\"\n",
    "    SELECT feature_id, coverage_ratio\n",
    "    FROM variant_cross_coverage\n",
    "    WHERE genome_id = 'B1A'\n",
    "    \"\"\",\n",
    "    index_col=['feature_id'],\n",
    "    con=con).squeeze()\n",
    "\n",
    "_min = -3.5\n",
    "_max = 1\n",
    "bins = np.logspace(_min, _max, num=int((_max - _min) * 10 + 1))\n",
    "plt.hist(cvrg_ratio_B1A + bins[0], bins=bins)\n",
    "plt.axvline(1/5, lw=1, linestyle='--', color='k')\n",
    "\n",
    "plt.xscale('log')\n",
    "#plt.yscale('symlog')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cvrg_ratio_B1B = pd.read_sql(\n",
    "    \"\"\"\n",
    "    SELECT feature_id, coverage_ratio\n",
    "    FROM variant_cross_coverage\n",
    "    WHERE genome_id = 'B1B'\n",
    "    \"\"\",\n",
    "    index_col=['feature_id'],\n",
    "    con=con).squeeze()\n",
    "\n",
    "_min = -3.5\n",
    "_max = 1\n",
    "bins = np.logspace(_min, _max, num=int((_max - _min) * 10 + 1))\n",
    "plt.hist(cvrg_ratio_B1B + bins[0], bins=bins)\n",
    "plt.axvline(1/5, lw=1, linestyle='--', color='k')\n",
    "plt.xscale('log')\n",
    "#plt.yscale('symlog')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "feature = pd.read_sql(\n",
    "    \"\"\"\n",
    "    SELECT feature_id, sequence_id, genome_id\n",
    "    FROM feature\n",
    "    JOIN sequence USING (sequence_id)\n",
    "    WHERE genome_id IN ('B1A', 'B1B')\n",
    "    \"\"\",\n",
    "    index_col=['feature_id'],\n",
    "    con=con\n",
    ")\n",
    "feature['sequence_alternating'] = (\n",
    "    feature.sequence_id.map(\n",
    "        pd.DataFrame({'sequence_id': feature.sequence_id.unique()})\n",
    "        .reset_index().set_index('sequence_id')\n",
    "        .squeeze().apply(lambda x: x % 2)\n",
    "    )\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cvrg_B1A = (\n",
    "    pd.read_sql(\n",
    "        \"\"\"\n",
    "        SELECT feature_id, library_id, coverage\n",
    "        FROM feature_library_coverage\n",
    "        JOIN feature USING (feature_id)\n",
    "        JOIN sequence USING (sequence_id)\n",
    "        WHERE genome_id = 'B1A'\n",
    "        \"\"\",\n",
    "    index_col=['feature_id', 'library_id'],\n",
    "    con=con)\n",
    "            .squeeze()\n",
    "            .unstack(fill_value=0)\n",
    "           )\n",
    "\n",
    "cvrg_B1B = (\n",
    "    pd.read_sql(\n",
    "        \"\"\"\n",
    "        SELECT feature_id, library_id, coverage\n",
    "        FROM feature_library_coverage\n",
    "        JOIN feature USING (feature_id)\n",
    "        JOIN sequence USING (sequence_id)\n",
    "        WHERE genome_id = 'B1B'\n",
    "        \"\"\",\n",
    "    index_col=['feature_id', 'library_id'],\n",
    "    con=con)\n",
    "            .squeeze()\n",
    "            .unstack(fill_value=0)\n",
    "           )\n",
    "\n",
    "\n",
    "\n",
    "with open('data/core.a.mags/B1A.g.library.list') as f:\n",
    "    library_list_B1A = list(set([line.strip() for line in f]) & set(cvrg_B1A.columns))\n",
    "    \n",
    "with open('data/core.a.mags/B1B.g.library.list') as f:\n",
    "    library_list_B1B = list(set([line.strip() for line in f]) & set(cvrg_B1B.columns))\n",
    "    \n",
    "assert not set(library_list_B1A) & set(library_list_B1B)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "library = pd.read_sql(\n",
    "    \"\"\"\n",
    "    SELECT *\n",
    "    FROM library\n",
    "    JOIN extraction USING (extraction_id)\n",
    "    JOIN sample USING (sample_id)\n",
    "    JOIN mouse USING (mouse_id)\n",
    "    \"\"\",\n",
    "    index_col='library_id',\n",
    "    con=con\n",
    ")\n",
    "library.loc['variant'] = 'untrusted'\n",
    "library.loc[library_list_B1A, 'variant'] = 'B1A'\n",
    "library.loc[library_list_B1B, 'variant'] = 'B1B'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Styles\n",
    "\n",
    "gene_content_heatmap_kwargs = dict(cmap='copper', norm=mpl.colors.SymLogNorm(1e-1, vmin=0, vmax=10, base=10))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_library_list = list(set(library_list_B1B) & set(cvrg_B1A.columns))\n",
    "\n",
    "cvrg_B1A_features_in_B1B_libs = cvrg_B1A[_library_list]\n",
    "median_cvrg_B1A_in_B1B_libs = cvrg_B1A_features_in_B1B_libs.median()\n",
    "ratio_cvrg_B1A_in_B1B_libs = cvrg_B1A_features_in_B1B_libs / median_cvrg_B1A_in_B1B_libs\n",
    "\n",
    "sns.heatmap(ratio_cvrg_B1A_in_B1B_libs.T, **gene_content_heatmap_kwargs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_library_list = list(set(library_list_B1A) & set(cvrg_B1A.columns))\n",
    "\n",
    "cvrg_B1A_features_in_B1A_libs = cvrg_B1A[_library_list]\n",
    "median_cvrg_B1A_in_B1A_libs = cvrg_B1A_features_in_B1A_libs.median()\n",
    "ratio_cvrg_B1A_in_B1A_libs = cvrg_B1A_features_in_B1A_libs / median_cvrg_B1A_in_B1A_libs\n",
    "\n",
    "sns.heatmap(ratio_cvrg_B1A_in_B1A_libs.T, **gene_content_heatmap_kwargs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_library_list = list(set(library_list_B1A) & set(cvrg_B1B.columns))\n",
    "\n",
    "cvrg_B1B_features_in_B1A_libs = cvrg_B1B[_library_list]\n",
    "median_cvrg_B1B_in_B1A_libs = cvrg_B1B_features_in_B1A_libs.median()\n",
    "ratio_cvrg_B1B_in_B1A_libs = cvrg_B1B_features_in_B1A_libs / median_cvrg_B1B_in_B1A_libs\n",
    "\n",
    "sns.heatmap(ratio_cvrg_B1B_in_B1A_libs.T, **gene_content_heatmap_kwargs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_library_list = list(set(library_list_B1B) & set(cvrg_B1B.columns))\n",
    "\n",
    "cvrg_B1B_features_in_B1B_libs = cvrg_B1B[_library_list]\n",
    "median_cvrg_B1B_in_B1B_libs = cvrg_B1B_features_in_B1B_libs.median()\n",
    "ratio_cvrg_B1B_in_B1B_libs = cvrg_B1B_features_in_B1B_libs / median_cvrg_B1B_in_B1B_libs\n",
    "\n",
    "sns.heatmap(ratio_cvrg_B1B_in_B1B_libs.T, **gene_content_heatmap_kwargs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cvrg_B1A_features = cvrg_B1A\n",
    "median_cvrg_B1A = cvrg_B1A_features.median()\n",
    "ratio_cvrg_B1A = cvrg_B1A_features / median_cvrg_B1A\n",
    "\n",
    "sns.clustermap(ratio_cvrg_B1A.T, row_cluster=False, col_cluster=False, **gene_content_heatmap_kwargs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cvrg_B1B_features = cvrg_B1B\n",
    "median_cvrg_B1B = cvrg_B1B_features.median()\n",
    "ratio_cvrg_B1B = cvrg_B1B_features / median_cvrg_B1B\n",
    "\n",
    "sns.clustermap(ratio_cvrg_B1B.T, row_cluster=False, col_cluster=False, **gene_content_heatmap_kwargs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dmat_raw = [[ratio_cvrg_B1A_in_B1A_libs, ratio_cvrg_B1B_in_B1A_libs],\n",
    "        [ratio_cvrg_B1A_in_B1B_libs, ratio_cvrg_B1B_in_B1B_libs]]\n",
    "xsplit, ysplit = dmat_raw[0][0].shape\n",
    "dmat = pd.concat([pd.concat([d for d in d1], axis='index') for d1 in dmat_raw], axis='columns')\n",
    "\n",
    "# Discretize a continuous colormap.\n",
    "ncolors = 9\n",
    "cmap = plt.cm.pink  # define the colormap\n",
    "# extract all colors from the .jet map\n",
    "cmaplist = [cmap(i) for i in np.linspace(0, 0.9, num=ncolors)]\n",
    "# force the first color entry to be grey\n",
    "\n",
    "# create the new map\n",
    "cmap = mpl.colors.LinearSegmentedColormap.from_list(\n",
    "    'Custom cmap', cmaplist, ncolors)\n",
    "\n",
    "#fig, axs = plt.subplots(2, 2, figsize=(10, 10))\n",
    "cg = sns.clustermap(\n",
    "    dmat.T, xticklabels=0, yticklabels=0,\n",
    "    row_colors=pd.DataFrame([library.site.map(color_map), library.variant.map(color_map)]).T,\n",
    "    col_colors=pd.DataFrame({'MAG': feature.genome_id.map(color_map), 'contig': feature.sequence_alternating.map(color_map)}),\n",
    "    row_cluster=True, col_cluster=False, metric='cosine',\n",
    "    cmap=cmap, vmax=1e1, norm=mpl.colors.SymLogNorm(1e-1, linscale=0.23, vmin=0, vmax=1e1, base=10),\n",
    "    rasterized=True, linewidths=0.,\n",
    "    cbar_kws=dict(orientation='horizontal', label='relative coverage'),\n",
    "    cbar_pos=(0.25, 0.02, 0.6, 0.03),\n",
    "    dendrogram_ratio=0.1,\n",
    "    figsize=(12.5, 8),\n",
    ")\n",
    "ax = cg.ax_heatmap\n",
    "ax.axvline(xsplit, color='gainsboro', lw=5)\n",
    "cg.ax_col_colors.axvline(xsplit, color='gainsboro', lw=5)\n",
    "# ax.axhline(ysplit, color='w')\n",
    "ax.set_xlabel('')\n",
    "ax.set_ylabel('')\n",
    "#ax.set_yticklabels([])\n",
    "\n",
    "col_colors_y_offset = 0.5\n",
    "cg.ax_col_colors.annotate(\n",
    "    'B1A Genes',\n",
    "    xy=(dmat_raw[0][0].shape[0] / 2, col_colors_y_offset),\n",
    "    va='center', ha='center'\n",
    ")\n",
    "cg.ax_col_colors.annotate(\n",
    "    'B1B Genes',\n",
    "    xy=(dmat_raw[0][0].shape[0] + dmat_raw[0][1].shape[0] / 2, col_colors_y_offset),\n",
    "    va='center', ha='center'\n",
    ")\n",
    "\n",
    "row_colors_x_offset = 1.6\n",
    "cg.ax_row_colors.annotate(\n",
    "    'B1A Libs.',\n",
    "    xy=(row_colors_x_offset, dmat_raw[0][0].shape[1] / 2),\n",
    "    va='center', ha='center', rotation=90,\n",
    ")\n",
    "cg.ax_row_colors.annotate(\n",
    "    'B1B Libs.',\n",
    "    xy=(row_colors_x_offset, dmat_raw[0][1].shape[1] + dmat_raw[1][1].shape[1] / 2),\n",
    "    va='center', ha='center', rotation=90,\n",
    ")\n",
    "\n",
    "plt.savefig('fig/B1_strain_comparison_heatmap_no_ambig.pdf', dpi=250, bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dmat_raw = [[ratio_cvrg_B1A, ratio_cvrg_B1B]]\n",
    "xsplit, ysplit = dmat_raw[0][0].shape\n",
    "dmat = pd.concat([pd.concat([d for d in d1], axis='index') for d1 in dmat_raw], axis='columns')\n",
    "# Discretize a continuous colormap.\n",
    "ncolors = 9\n",
    "cmap = plt.cm.pink  # define the colormap\n",
    "# extract all colors from the .jet map\n",
    "cmaplist = [cmap(i) for i in np.linspace(0, 0.9, num=ncolors)]\n",
    "# force the first color entry to be grey\n",
    "\n",
    "# create the new map\n",
    "cmap = mpl.colors.LinearSegmentedColormap.from_list(\n",
    "    'Custom cmap', cmaplist, ncolors)\n",
    "\n",
    "#fig, axs = plt.subplots(2, 2, figsize=(10, 10))\n",
    "cg = sns.clustermap(\n",
    "    dmat.T, xticklabels=0, yticklabels=0,\n",
    "    row_colors=pd.DataFrame([library.site.map(color_map), library.variant.map(color_map)]).T,\n",
    "    col_colors=pd.DataFrame({'MAG': feature.genome_id.map(color_map), 'contig': feature.sequence_alternating.map(color_map)}),\n",
    "    row_cluster=True, col_cluster=False, metric='cosine',\n",
    "    cmap=cmap, vmax=1e1, norm=mpl.colors.SymLogNorm(1e-1, linscale=0.23, vmin=0, vmax=1e1, base=10),\n",
    "    rasterized=True, linewidths=0.,\n",
    "    cbar_kws=dict(orientation='horizontal', label='relative coverage'),\n",
    "    cbar_pos=(0.25, 0.02, 0.6, 0.03),\n",
    "    dendrogram_ratio=0.1,\n",
    "    figsize=(12.5, 8),\n",
    ")\n",
    "ax = cg.ax_heatmap\n",
    "ax.axvline(xsplit, color='gainsboro', lw=5)\n",
    "cg.ax_col_colors.axvline(xsplit, color='gainsboro', lw=5)\n",
    "# ax.axhline(ysplit, color='w')\n",
    "ax.set_xlabel('')\n",
    "ax.set_ylabel('')\n",
    "#ax.set_yticklabels([])\n",
    "\n",
    "col_colors_y_offset = 0.5\n",
    "cg.ax_col_colors.annotate(\n",
    "    'B1A Genes',\n",
    "    fontsize=8,\n",
    "    xy=(dmat_raw[0][0].shape[0] / 2, col_colors_y_offset),\n",
    "    va='center', ha='center'\n",
    ")\n",
    "cg.ax_col_colors.annotate(\n",
    "    'B1B Genes',\n",
    "    fontsize=8,\n",
    "    xy=(dmat_raw[0][0].shape[0] + dmat_raw[0][1].shape[0] / 2, col_colors_y_offset),\n",
    "    va='center', ha='center'\n",
    ")\n",
    "\n",
    "row_colors_x_offset = 1.6\n",
    "cg.ax_row_colors.annotate(\n",
    "    'B1A Libs.',\n",
    "    fontsize=8,\n",
    "    xy=(row_colors_x_offset, len(library_list_B1A) / 2),\n",
    "    va='center', ha='center', rotation=90,\n",
    ")\n",
    "cg.ax_row_colors.annotate(\n",
    "    'B1B Libs.',\n",
    "    fontsize=8,\n",
    "    xy=(row_colors_x_offset, len(library_list_B1A) + len(library_list_B1B) / 2),\n",
    "    va='center', ha='center', rotation=90,\n",
    ")\n",
    "\n",
    "plt.savefig('fig/B1_strain_comparison_heatmap.pdf', dpi=250, bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def list_unique_functions(filt, func_map):\n",
    "    return func_map[func_map.feature_id.isin(idxwhere(filt))].func_id.unique()\n",
    "\n",
    "def count_unique_functions(filt, func_map):\n",
    "    return len(list_unique_functions(filt, func_map))"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "JOIN feature_x_ko USING (feature_id)\n",
    "JOIN feature_to_cog USING (feature_id)\n",
    "JOIN feature_x_cazy_domain USING (feature_id)\n",
    "JOIN feature_x_pfam_domain USING (feature_id)\n",
    "JOIN feature_to_opf USING (feature_id)\n",
    "JOIN feature_x_tigr_domain USING (feature_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_feature_x_func = pd.read_sql(\n",
    "    \"\"\"\n",
    "    SELECT feature_id, feature_id AS func_id\n",
    "    FROM feature\n",
    "    JOIN sequence USING (sequence_id)\n",
    "    WHERE genome_id IN ('B1A', 'B1B')\n",
    "      AND func_id NOT NULL\n",
    "    \"\"\",\n",
    "    con=con\n",
    ")\n",
    "\n",
    "((count_unique_functions(cvrg_ratio_B1A > -1, _feature_x_func),\n",
    "  count_unique_functions(cvrg_ratio_B1A < 0.2, _feature_x_func)),\n",
    " (count_unique_functions(cvrg_ratio_B1B > -1, _feature_x_func),\n",
    "  count_unique_functions(cvrg_ratio_B1B < 0.2, _feature_x_func)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_feature_x_func = pd.read_sql(\n",
    "    \"\"\"\n",
    "    SELECT feature_id, ko_id AS func_id\n",
    "    FROM feature\n",
    "    JOIN sequence USING (sequence_id)\n",
    "    JOIN feature_x_ko USING (feature_id)\n",
    "    WHERE genome_id IN ('B1A', 'B1B')\n",
    "      AND func_id NOT NULL\n",
    "    \"\"\",\n",
    "    con=con\n",
    ")\n",
    "\n",
    "((count_unique_functions(cvrg_ratio_B1A > -1, _feature_x_func),\n",
    "  count_unique_functions(cvrg_ratio_B1A < 0.2, _feature_x_func)),\n",
    " (count_unique_functions(cvrg_ratio_B1B > -1, _feature_x_func),\n",
    "  count_unique_functions(cvrg_ratio_B1B < 0.2, _feature_x_func)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_feature_x_func = pd.read_sql(\n",
    "    \"\"\"\n",
    "    SELECT feature_id, cog_id AS func_id\n",
    "    FROM feature\n",
    "    JOIN sequence USING (sequence_id)\n",
    "    JOIN feature_to_cog USING (feature_id)\n",
    "    WHERE genome_id IN ('B1A', 'B1B')\n",
    "      AND func_id NOT NULL\n",
    "    \"\"\",\n",
    "    con=con\n",
    ")\n",
    "\n",
    "((count_unique_functions(cvrg_ratio_B1A > -1, _feature_x_func),\n",
    "  count_unique_functions(cvrg_ratio_B1A < 0.2, _feature_x_func)),\n",
    " (count_unique_functions(cvrg_ratio_B1B > -1, _feature_x_func),\n",
    "  count_unique_functions(cvrg_ratio_B1B < 0.2, _feature_x_func)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_feature_x_func = pd.read_sql(\n",
    "    \"\"\"\n",
    "    SELECT feature_id, domain_id AS func_id\n",
    "    FROM feature\n",
    "    JOIN sequence USING (sequence_id)\n",
    "    JOIN feature_x_cazy_domain USING (feature_id)\n",
    "    WHERE genome_id IN ('B1A', 'B1B')\n",
    "      AND func_id NOT NULL\n",
    "    \"\"\",\n",
    "    con=con\n",
    ")\n",
    "\n",
    "((count_unique_functions(cvrg_ratio_B1A > -1, _feature_x_func),\n",
    "  count_unique_functions(cvrg_ratio_B1A < 0.2, _feature_x_func)),\n",
    " (count_unique_functions(cvrg_ratio_B1B > -1, _feature_x_func),\n",
    "  count_unique_functions(cvrg_ratio_B1B < 0.2, _feature_x_func)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_feature_x_func = pd.read_sql(\n",
    "    \"\"\"\n",
    "    SELECT feature_id, domain_id AS func_id\n",
    "    FROM feature\n",
    "    JOIN sequence USING (sequence_id)\n",
    "    JOIN feature_x_pfam_domain USING (feature_id)\n",
    "    WHERE genome_id IN ('B1A', 'B1B')\n",
    "      AND func_id NOT NULL\n",
    "    \"\"\",\n",
    "    con=con\n",
    ")\n",
    "\n",
    "((count_unique_functions(cvrg_ratio_B1A > -1, _feature_x_func),\n",
    "  count_unique_functions(cvrg_ratio_B1A < 0.2, _feature_x_func)),\n",
    " (count_unique_functions(cvrg_ratio_B1B > -1, _feature_x_func),\n",
    "  count_unique_functions(cvrg_ratio_B1B < 0.2, _feature_x_func)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_feature_x_func = pd.read_sql(\n",
    "    \"\"\"\n",
    "    SELECT feature_id, opf_id AS func_id\n",
    "    FROM feature\n",
    "    JOIN sequence USING (sequence_id)\n",
    "    JOIN feature_to_opf USING (feature_id)\n",
    "    WHERE genome_id IN ('B1A', 'B1B')\n",
    "      AND func_id NOT NULL\n",
    "    \"\"\",\n",
    "    con=con\n",
    ")\n",
    "\n",
    "((count_unique_functions(cvrg_ratio_B1A > -1, _feature_x_func),\n",
    "  count_unique_functions(cvrg_ratio_B1A < 0.2, _feature_x_func)),\n",
    " (count_unique_functions(cvrg_ratio_B1B > -1, _feature_x_func),\n",
    "  count_unique_functions(cvrg_ratio_B1B < 0.2, _feature_x_func)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_feature_x_func = pd.read_sql(\n",
    "    \"\"\"\n",
    "    SELECT feature_id, domain_id AS func_id\n",
    "    FROM feature\n",
    "    JOIN sequence USING (sequence_id)\n",
    "    JOIN feature_x_tigr_domain USING (feature_id)\n",
    "    WHERE genome_id IN ('B1A', 'B1B')\n",
    "      AND func_id NOT NULL\n",
    "    \"\"\",\n",
    "    con=con\n",
    ")\n",
    "\n",
    "((count_unique_functions(cvrg_ratio_B1A > -1, _feature_x_func),\n",
    "  count_unique_functions(cvrg_ratio_B1A < 0.2, _feature_x_func)),\n",
    " (count_unique_functions(cvrg_ratio_B1B > -1, _feature_x_func),\n",
    "  count_unique_functions(cvrg_ratio_B1B < 0.2, _feature_x_func)))\n",
    "\n",
    "_description = pd.read_sql(\n",
    "    \"\"\"\n",
    "    SELECT domain_id, tigr_description\n",
    "    FROM tigr_domain;\n",
    "    \"\"\",\n",
    "    index_col='domain_id',\n",
    "    con=con,\n",
    ")\n",
    "\n",
    "_description.loc[list_unique_functions(cvrg_ratio_B1A < 0.2, _feature_x_func)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mean_normalized_cvrg_B1A = cvrg_B1A[library_list_B1A].apply(lambda x: x / x.mean()).mean(axis=1)\n",
    "mean_normalized_cvrg_B1B = cvrg_B1B[library_list_B1B].apply(lambda x: x / x.mean()).mean(axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.scatter(mean_normalized_cvrg_B1A, cvrg_ratio_B1A, s=5, alpha=0.5)\n",
    "plt.xscale('log')\n",
    "plt.yscale('log')\n",
    "plt.ylim(1e-6, 1e3)\n",
    "plt.axhline(1, lw=1, linestyle='--', color='k', zorder=0)\n",
    "plt.axhline(1/5, lw=1, linestyle=':', color='k', zorder=0)\n",
    "plt.axvline(1, lw=1, linestyle='--', color='k', zorder=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "(np.abs(np.log2(mean_normalized_cvrg_B1A)) > 1).mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.scatter(mean_normalized_cvrg_B1B, cvrg_ratio_B1B, s=5, alpha=0.5)\n",
    "plt.xscale('log')\n",
    "plt.yscale('log')\n",
    "plt.ylim(1e-6, 1e3)\n",
    "plt.axhline(1, lw=1, linestyle='--', color='k', zorder=0)\n",
    "plt.axhline(1/5, lw=1, linestyle=':', color='k', zorder=0)\n",
    "plt.axvline(1, lw=1, linestyle='--', color='k', zorder=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "(np.abs(np.log2(mean_normalized_cvrg_B1B)) > 1).mean()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {
    "height": "205px",
    "width": "160px"
   },
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}