{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from matplotlib import pyplot as plt\n",
    "from Bio import Phylo\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from copy import deepcopy\n",
    "from io import StringIO\n",
    "import matplotlib as mpl\n",
    "from scripts.lib.plotting import load_style\n",
    "import sqlite3\n",
    "from itertools import product\n",
    "from sklearn.manifold import MDS\n",
    "from collections import defaultdict\n",
    "\n",
    "\n",
    "loaded_style = load_style('paper')\n",
    "savefig = loaded_style['savefig']\n",
    "fullwidth = loaded_style['fullwidth']\n",
    "halfwidth = loaded_style['halfwidth']\n",
    "\n",
    "def annotate_clade_with_plotting_metadata(clade, left=0, bottom=0):\n",
    "    branch_length = clade.branch_length\n",
    "    if not branch_length:\n",
    "        branch_length = 0\n",
    "    clade.left = left\n",
    "    clade.right = left + branch_length\n",
    "    if clade.is_terminal():\n",
    "        clade.y = bottom\n",
    "        return 0, clade.y  # total_offset, root_pos\n",
    "    else:\n",
    "        roots = []\n",
    "        total_offset = 0\n",
    "        for c in clade.clades:\n",
    "            offset, root_pos = annotate_clade_with_plotting_metadata(\n",
    "                    c, left=clade.right, bottom=bottom + total_offset)\n",
    "            roots.append(root_pos)\n",
    "            total_offset += offset + 1\n",
    "        clade.y = (min(roots) + max(roots)) / 2.0\n",
    "        return total_offset - 1, clade.y\n",
    "\n",
    "def plot_clade(clade, ax,\n",
    "               tax_plotter=lambda c, ax: (),\n",
    "               node_plotter=lambda c, ax: (),\n",
    "               lw=1):\n",
    "    ax.hlines(clade.y, clade.left, clade.right, lw=lw)\n",
    "    children_ys = [plot_clade(c, ax,\n",
    "                              tax_plotter=tax_plotter,\n",
    "                              node_plotter=node_plotter,\n",
    "                             )\n",
    "                   for c in clade.clades]\n",
    "    if children_ys:\n",
    "        ymin = min(children_ys)\n",
    "        ymax = max(children_ys)\n",
    "        ax.vlines(clade.right, ymin, ymax, lw=lw)\n",
    "    if clade.is_terminal():\n",
    "        tax_plotter(clade, ax)\n",
    "    node_plotter(clade, ax)\n",
    "    return clade.y\n",
    "\n",
    "def plot_tree(tree, ax, **kwargs):\n",
    "    tree = deepcopy(tree)\n",
    "    clade = getattr(tree, 'clade', tree)\n",
    "    annotate_clade_with_plotting_metadata(clade)\n",
    "    plot_clade(clade, ax, **kwargs)\n",
    "    return tree\n",
    "    \n",
    "def fold_by_group(clade, key):\n",
    "    if clade.is_terminal():\n",
    "        clade.group = key(clade)\n",
    "        clade.minbranch = 0\n",
    "        clade.maxbranch = 0\n",
    "        clade.count = 1\n",
    "        return clade.group, clade.branch_length, clade.branch_length, 1\n",
    "    else:\n",
    "        child_group, child_minbranch, child_maxbranch, child_count = \\\n",
    "                zip(*[fold_by_group(c, key) for c in clade.clades])\n",
    "        unique_groups = set(child_group)\n",
    "        if (None in unique_groups) or len(unique_groups) > 1:\n",
    "            # Children are not of one group\n",
    "            # don't collapse anything and pass on null values\n",
    "            return None, 0, 0, None\n",
    "        else:\n",
    "            # It IS a unique group\n",
    "            # so drop the children\n",
    "            clade.count = sum(child_count)\n",
    "            clade.clades = []\n",
    "            # and pass on the group and the branch lengths\n",
    "            clade.group = unique_groups.pop()\n",
    "            assert not unique_groups, \"After popping, I expect unique_groups to be an empty set.\"\n",
    "            clade.minbranch = min(child_minbranch)\n",
    "            clade.maxbranch = max(child_maxbranch)\n",
    "            return (clade.group,\n",
    "                    clade.minbranch + clade.branch_length,\n",
    "                    clade.maxbranch + clade.branch_length,\n",
    "                    clade.count)"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "def _tax_plotter(clade, ax):\n",
    "    if clade.name:\n",
    "        label = clade.name\n",
    "    else:\n",
    "        label = '{} ({})'.format(clade.group, clade.count)\n",
    "    if hasattr(clade, 'minbranch'):\n",
    "        x = [clade.right,\n",
    "             clade.right + clade.minbranch,\n",
    "             clade.right + clade.maxbranch,\n",
    "             clade.right\n",
    "            ]\n",
    "        y = [clade.y + 0.45] * 2 + [clade.y - 0.45] * 2\n",
    "        ax.add_patch(plt.Polygon(xy=list(zip(x, y)), alpha=0.25))\n",
    "    ax.annotate(label, xy=(clade.right + 0.0075, clade.y), ha='left', va='center', weight='bold')\n",
    "\n",
    "test_tree_string = '((D:0.723274,((F:0.567784,Z:0.3):0.3, Q:0.4):0.3)1.000000:0.167192,(B:0.279326,H:0.756049)1.000000:0.807788);'\n",
    "tree = Phylo.read(StringIO(test_tree_string), 'newick')\n",
    "fig, ax = plt.subplots()\n",
    "plot_tree(tree, ax, tax_plotter=_tax_plotter)\n",
    "\n",
    "\n",
    "fold_by_group(tree.clade, lambda c: {'H': 'Group 0', 'B': 'Group 0', 'Q': 'Group 1', 'Z': 'Group 1', 'F': 'Group 1', 'D': 'Group 0'}[c.name])\n",
    "fig, ax = plt.subplots()\n",
    "plot_tree(tree, ax, tax_plotter=_tax_plotter)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "color_map = { 'starch': 'blue'\n",
    "            , 'host': 'purple'\n",
    "            , 'plant': 'green'\n",
    "            }\n",
    "\n",
    "italic_list = [ 'Barnesiella_viscericola_DSM_18177'\n",
    "              , 'Barnesiella_intestinihominis_YIT_11860'\n",
    "              , 'Bacteroides_ovatus_ATCC_8483'\n",
    "              , 'Bacteroides_thetaiotaomicron_VPI5482'\n",
    "              , 'Porphyromonas_gingivalis_ATCC_33277'\n",
    "              , 'Homeothermus_arabinoxylanisolvens'\n",
    "              , 'Muribaculum_intestinale_DSM_28989T'\n",
    "              , 'Duncaniella_muris_DSM_103720T'\n",
    "              , 'Duncaniella_freteri_DSM_108168T'\n",
    "              , 'Duncaniella_dubosii_DSM_107170T'\n",
    "              , 'Paramuribaculum_intestinale_DSM_100749T'\n",
    "              , 'Amulumruptor_caecigallinarius'\n",
    "              ]\n",
    "\n",
    "con = sqlite3.connect('data/core.muri2.2.denorm.db')\n",
    "\n",
    "mag = (\n",
    "    pd.read_sql(\n",
    "    \"\"\"\n",
    "SELECT *\n",
    "FROM genome\n",
    "JOIN checkm USING (genome_id)\n",
    "JOIN genome_group USING (genome_id)\n",
    "WHERE genome_group_id = 'muri2'\n",
    "    \"\"\",\n",
    "    index_col='genome_id',\n",
    "    con=con,\n",
    "    )\n",
    ")\n",
    "\n",
    "rename_taxon = {\n",
    "#                  'Barnesiella_viscericola_DSM_18177': 'Bv'\n",
    "#                , 'Bacteroides_ovatus_ATCC_8483': 'Bo'\n",
    "#                , 'Bacteroides_thetaiotaomicron_VPI5482': 'Bt'\n",
    "#                , 'Porphyromonas_gingivalis_ATCC_33277': 'Pg'\n",
    "#                , 'Homeothermus_arabinoxylanisolvens': 'Ha'\n",
    "#                , 'Muribaculum_intestinale_DSM_28989T': 'Mi'\n",
    "#                , 'Duncaniella_muris_DSM_103720T': 'Dm'\n",
    "#                , 'Duncaniella_freteri_DSM_108168T': 'Df'\n",
    "#                , 'Duncaniella_dubosii_DSM_107170T': 'Dd'\n",
    "#                , 'Paramuribaculum_intestinale_DSM_100749T': 'Pi'\n",
    "#                , 'Amulumruptor_caecigallinarius': 'Ac'\n",
    "#                , 'Barnesiella_intestinihominis_YIT_11860': 'Bi'\n",
    "               }\n",
    "\n",
    "for _name, x in mag.iterrows():\n",
    "    if _name.startswith('GCA_'):\n",
    "        rename_taxon[_name] = mag.loc[_name].genbank_accession\n",
    "    else:\n",
    "        rename_taxon[_name] = _name.replace('_', ' ')\n",
    "\n",
    "\n",
    "colormap = color_map = { 'ormerod-starch': 'blue'\n",
    "                       , 'ormerod-host': 'purple'\n",
    "                       , 'ormerod-plant': 'green'\n",
    "                       , 'public': 'grey'\n",
    "                       , 'here': 'black'\n",
    "                       , 'outgroup': 'lightgrey'\n",
    "                       }\n",
    "\n",
    "marker_map = {'ormerod-plant': 'o', 'ormerod-host': 'o', 'ormerod-starch': 'o', 'public': 'o', 'here': '^', 'outgroup': 'o'}\n",
    "\n",
    "def _tax_plotter(clade, ax):\n",
    "    if clade.name in rename_taxon:\n",
    "        # Check if there's a renaming for a given clade.\n",
    "        label = rename_taxon[clade.name]\n",
    "    else:\n",
    "        label = clade.name\n",
    "        \n",
    "#     if mag.subgroup[clade.name] == 'here':\n",
    "#         label = label + '*'\n",
    "        \n",
    "    if clade.name in italic_list:\n",
    "        textstyle = 'italic'\n",
    "    else:\n",
    "        textstyle = 'normal'\n",
    "    \n",
    "    color = colormap[mag.subgroup[clade.name]]\n",
    "    ax.scatter(\n",
    "        [clade.right + 0.01],\n",
    "        [clade.y],\n",
    "        color=color,\n",
    "        marker=marker_map[mag.subgroup[clade.name]],\n",
    "        s=5,\n",
    "    )\n",
    "    ax.annotate(label, xy=(clade.right + 0.02, clade.y),\n",
    "                ha='left', va='center',\n",
    "                weight='bold', style=textstyle,\n",
    "                color=color, fontsize=6.5)\n",
    "    \n",
    "def _node_plotter(clade, ax):\n",
    "    confidence = clade.confidence\n",
    "    if confidence is None:\n",
    "        return\n",
    "\n",
    "    if confidence > 0.95:\n",
    "        if clade.is_terminal():\n",
    "            hpos = clade.right - 0.013\n",
    "        else:\n",
    "            hpos = clade.right\n",
    "        color = 'k'\n",
    "        ax.scatter([hpos], [clade.y], color=color,\n",
    "                   marker='o', s=5, linewidths=0, edgecolors='k', zorder=10)\n",
    "\n",
    "    \n",
    "    \n",
    "tree = Phylo.read('data/core.a.mags.muri2.g.final.marker_genes.gb.prot.nwk', 'newick')\n",
    "outgroup = [\n",
    "              'Porphyromonas_gingivalis_ATCC_33277'\n",
    "#             , 'Bacteroides_ovatus_ATCC_8483'\n",
    "#             , 'Bacteroides_thetaiotaomicron_VPI5482'\n",
    "#            , 'Barnesiella_viscericola_DSM_18177'\n",
    "           ]\n",
    "view_outgroup = [\n",
    "           ]\n",
    "\n",
    "tree.root_with_outgroup(outgroup)\n",
    "for t in outgroup:\n",
    "    if t not in view_outgroup:\n",
    "        tree.prune(t)\n",
    "        \n",
    "tree.clade.branch_length = 0.06\n",
    "#tree.clade.confidence = None\n",
    "tree.collapse_all(lambda c: c.confidence is not None and c.confidence < 0.7)\n",
    "#tree.ladderize(reverse=True)\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(fullwidth, 18))\n",
    "tree_annotated = plot_tree(tree, ax,\n",
    "                           tax_plotter=_tax_plotter,\n",
    "                           node_plotter=_node_plotter)\n",
    "\n",
    "#for confidence_label, color in [('> 95%', 'black'),\n",
    "#                                ('> 90%', 'grey'),\n",
    "#                                ('> 80%', 'lightgrey'),\n",
    "#                                ('< 80%', 'white')\n",
    "#                                ('> 80%', 'white')\n",
    "#                               ]\n",
    "#    ax.scatter([], [], color=color, label=confidence_label, s=50, linewidths=2, edgecolors='k')\n",
    "ax.set_yticks([])\n",
    "ax.set_xticks([])\n",
    "#ax.legend(loc='upper left', title='confidence')\n",
    "\n",
    "scale_xy = (0.1, 5.5)\n",
    "scale_length = 0.2\n",
    "_x, _y = scale_xy\n",
    "ax.hlines([_y], [_x], [_x + scale_length], lw=1)\n",
    "ax.annotate('{:.1}'.format(scale_length),\n",
    "            xy=(_x + scale_length / 2, _y + 0.4),\n",
    "            ha='center', fontsize=8)\n",
    "\n",
    "# panel = 'A'\n",
    "# ax.annotate(panel, xy=(0.02, 1.02), xycoords='axes fraction', fontsize=12)\n",
    "\n",
    "\n",
    "ax.axis('off')\n",
    "#fig.tight_layout()\n",
    " \n",
    "plt.savefig('fig/phylogenetics_concatenated.pdf', dpi=250, bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "color_map = { 'starch': 'blue'\n",
    "            , 'host': 'purple'\n",
    "            , 'plant': 'green'\n",
    "            }\n",
    "\n",
    "italic_list = [ 'Barnesiella_viscericola_DSM_18177'\n",
    "              , 'Bacteroides_ovatus_ATCC_8483'\n",
    "              , 'Bacteroides_thetaiotaomicron_VPI5482'\n",
    "              , 'Porphyromonas_gingivalis_ATCC_33277'\n",
    "              , 'Homeothermus_arabinoxylanisolvens'\n",
    "              , 'Muribaculum_intestinale_DSM_28989T'\n",
    "              , 'Duncaniella_muris_DSM_103720T'\n",
    "              , 'Duncaniella_freteri_DSM_108168T'\n",
    "              , 'Duncaniella_dubosii_DSM_107170T'\n",
    "              , 'Paramuribaculum_intestinale_DSM_100749T'\n",
    "              , 'Amulumruptor_caecigallinarius'\n",
    "              , 'Barnesiella_intestinihominis_YIT_11860'\n",
    "              ]\n",
    "\n",
    "con = sqlite3.connect('data/core.muri2.2.denorm.db')\n",
    "\n",
    "mag = (\n",
    "    pd.read_sql(\n",
    "    \"\"\"\n",
    "SELECT *\n",
    "FROM genome\n",
    "JOIN checkm USING (genome_id)\n",
    "JOIN genome_group USING (genome_id)\n",
    "WHERE genome_group_id = 'muri2'\n",
    "    \"\"\",\n",
    "    index_col='genome_id',\n",
    "    con=con,\n",
    "    )\n",
    ")\n",
    "\n",
    "rename_taxon = {\n",
    "#                  'Barnesiella_viscericola_DSM_18177': 'Bv'\n",
    "#                , 'Bacteroides_ovatus_ATCC_8483': 'Bo'\n",
    "#                , 'Bacteroides_thetaiotaomicron_VPI5482': 'Bt'\n",
    "#                , 'Porphyromonas_gingivalis_ATCC_33277': 'Pg'\n",
    "#                , 'Homeothermus_arabinoxylanisolvens': 'Ha'\n",
    "#                , 'Muribaculum_intestinale_DSM_28989T': 'Mi'\n",
    "#                , 'Duncaniella_muris_DSM_103720T': 'Dm'\n",
    "#                , 'Duncaniella_freteri_DSM_108168T': 'Df'\n",
    "#                , 'Duncaniella_dubosii_DSM_107170T': 'Dd'\n",
    "#                , 'Paramuribaculum_intestinale_DSM_100749T': 'Pi'\n",
    "#                , 'Amulumruptor_caecigallinarius': 'Ac'\n",
    "#                , 'Barnesiella_intestinihominis_YIT_11860': 'Bi'\n",
    "               }\n",
    "\n",
    "for _name, x in mag.iterrows():\n",
    "    if _name.startswith('GCA_'):\n",
    "        rename_taxon[_name] = mag.loc[_name].genbank_accession\n",
    "    else:\n",
    "        rename_taxon[_name] = _name.replace('_', ' ')\n",
    "\n",
    "\n",
    "colormap = color_map = { 'ormerod-starch': 'blue'\n",
    "                       , 'ormerod-host': 'purple'\n",
    "                       , 'ormerod-plant': 'green'\n",
    "                       , 'public': 'grey'\n",
    "                       , 'here': 'black'\n",
    "                       , 'outgroup': 'lightgrey'\n",
    "                       }\n",
    "\n",
    "marker_map = {'ormerod-plant': 'o', 'ormerod-host': 'o', 'ormerod-starch': 'o', 'public': 'o', 'here': '^', 'outgroup': 'o'}\n",
    "\n",
    "def _tax_plotter(clade, ax):\n",
    "    if clade.name in rename_taxon:\n",
    "        # Check if there's a renaming for a given clade.\n",
    "        label = rename_taxon[clade.name]\n",
    "    else:\n",
    "        label = clade.name\n",
    "        \n",
    "#     if mag.subgroup[clade.name] == 'here':\n",
    "#         label = label + '*'\n",
    "        \n",
    "    if clade.name in italic_list:\n",
    "        textstyle = 'italic'\n",
    "    else:\n",
    "        textstyle = 'normal'\n",
    "    \n",
    "    color = colormap[mag.subgroup[clade.name]]\n",
    "    ax.scatter(\n",
    "        [clade.right + 0.005],\n",
    "        [clade.y],\n",
    "        color=color,\n",
    "        marker=marker_map[mag.subgroup[clade.name]],\n",
    "        s=5,\n",
    "    )\n",
    "    ax.annotate(label, xy=(clade.right + 0.01, clade.y),\n",
    "                ha='left', va='center',\n",
    "                weight='bold', style=textstyle,\n",
    "                color=color, fontsize=6.5)\n",
    "    \n",
    "def _node_plotter(clade, ax):\n",
    "    confidence = clade.confidence\n",
    "    if confidence is None:\n",
    "        return\n",
    "\n",
    "    if confidence > 0.95:\n",
    "        if clade.is_terminal():\n",
    "            hpos = clade.right - 0.013\n",
    "        else:\n",
    "            hpos = clade.right\n",
    "        color = 'k'\n",
    "        ax.scatter([hpos], [clade.y], color=color,\n",
    "                   marker='o', s=5, linewidths=0, edgecolors='k', zorder=10)\n",
    "\n",
    "    \n",
    "    \n",
    "tree = Phylo.read('data/core.a.mags.muri2.g.final.cds.TIGR02013-hits.sqz.gb.prot.nwk', 'newick')\n",
    "# Rename taxa to species rather than gene.\n",
    "for t in tree.get_terminals():\n",
    "    t.name = t.name.rsplit('_', 1)[0]\n",
    "    \n",
    "outgroup = [\n",
    "#               'Porphyromonas_gingivalis_ATCC_33277'\n",
    "            'Bacteroides_ovatus_ATCC_8483'\n",
    "            , 'Bacteroides_thetaiotaomicron_VPI5482'\n",
    "           , 'Barnesiella_viscericola_DSM_18177'\n",
    "           ]\n",
    "view_outgroup = []\n",
    "\n",
    "tree.root_with_outgroup(outgroup)\n",
    "for t in outgroup:\n",
    "    if t not in view_outgroup:\n",
    "        tree.prune(t)\n",
    "        \n",
    "tree.clade.branch_length = 0.06\n",
    "#tree.clade.confidence = None\n",
    "tree.collapse_all(lambda c: c.confidence is not None and c.confidence < 0.7)\n",
    "#tree.ladderize(reverse=True)\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(halfwidth, 13))\n",
    "tree_annotated = plot_tree(tree, ax,\n",
    "                           tax_plotter=_tax_plotter,\n",
    "                           node_plotter=_node_plotter)\n",
    "\n",
    "#for confidence_label, color in [('> 95%', 'black'),\n",
    "#                                ('> 90%', 'grey'),\n",
    "#                                ('> 80%', 'lightgrey'),\n",
    "#                                ('< 80%', 'white')\n",
    "#                                ('> 80%', 'white')\n",
    "#                               ]\n",
    "#    ax.scatter([], [], color=color, label=confidence_label, s=50, linewidths=2, edgecolors='k')\n",
    "ax.set_yticks([])\n",
    "ax.set_xticks([])\n",
    "#ax.legend(loc='upper left', title='confidence')\n",
    "\n",
    "scale_xy = (0.02, 5.5)\n",
    "scale_length = 0.05\n",
    "_x, _y = scale_xy\n",
    "ax.hlines([_y], [_x], [_x + scale_length], lw=1)\n",
    "ax.annotate('{:.1}'.format(scale_length),\n",
    "            xy=(_x + scale_length / 2, _y + 0.4),\n",
    "            ha='center', fontsize=8)\n",
    "\n",
    "# panel = 'A'\n",
    "# ax.annotate(panel, xy=(0.02, 1.02), xycoords='axes fraction', fontsize=12)\n",
    "\n",
    "\n",
    "ax.axis('off')\n",
    "#fig.tight_layout()\n",
    " \n",
    "plt.savefig('fig/phylogenetics_rpoB.pdf', dpi=250, bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "rename_taxon = {\n",
    "                 'Barnesiella_viscericola_DSM_18177': 'Bv'\n",
    "               , 'Bacteroides_ovatus_ATCC_8483': 'Bo'\n",
    "               , 'Bacteroides_thetaiotaomicron_VPI5482': 'Bt'\n",
    "               , 'Porphyromonas_gingivalis_ATCC_33277': 'Pg'\n",
    "               , 'Homeothermus_arabinoxylanisolvens': 'Ha'\n",
    "               , 'Muribaculum_intestinale_DSM_28989T': 'Mi'\n",
    "               , 'Duncaniella_muris_DSM_103720T': 'Dm'\n",
    "               , 'Duncaniella_freteri_DSM_108168T': 'Df'\n",
    "               , 'Duncaniella_dubosii_DSM_107170T': 'Dd'\n",
    "               , 'Paramuribaculum_intestinale_DSM_100749T': 'Pi'\n",
    "               , 'Amulumruptor_caecigallinarius': 'Ac'\n",
    "               } \n",
    "                 \n",
    "color_map = { 'starch': 'blue'\n",
    "            , 'host': 'purple'\n",
    "            , 'plant': 'green'\n",
    "            }\n",
    "\n",
    "italic_list = [ 'Barnesiella_viscericola_DSM_18177'\n",
    "              , 'Bacteroides_ovatus_ATCC_8483'\n",
    "              , 'Bacteroides_thetaiotaomicron_VPI5482'\n",
    "              , 'Porphyromonas_gingivalis_ATCC_33277'\n",
    "              , 'Homeothermus_arabinoxylanisolvens'\n",
    "              , 'Muribaculum_intestinale_DSM_28989T'\n",
    "              , 'Duncaniella_muris_DSM_103720T'\n",
    "              , 'Duncaniella_freteri_DSM_108168T'\n",
    "              , 'Duncaniella_dubosii_DSM_107170T'\n",
    "              , 'Paramuribaculum_intestinale_DSM_100749T'\n",
    "              , 'Amulumruptor_caecigallinarius'\n",
    "              ]\n",
    "\n",
    "mag = pd.read_table('meta/genome.tsv', index_col='genome_id')\n",
    "\n",
    "def _tax_plotter(clade, ax):\n",
    "    if clade.name in rename_taxon:\n",
    "        # Check if there's a renaming for a given clade.\n",
    "        label = rename_taxon[clade.name]\n",
    "    else:\n",
    "        label = clade.name\n",
    "        \n",
    "    if mag.genome_type[clade.name] == 'here':\n",
    "        label = label + '*'\n",
    "        \n",
    "    if clade.name in italic_list:\n",
    "        textstyle = 'italic'\n",
    "    else:\n",
    "        textstyle = 'normal'\n",
    "    \n",
    "    guild = mag.ormerod_guild[clade.name]\n",
    "    if guild in color_map:\n",
    "        color = color_map[guild]\n",
    "    else:\n",
    "        color = 'black'\n",
    "    ax.annotate(label, xy=(clade.right + 0.015, clade.y),\n",
    "                ha='left', va='center',\n",
    "                weight='bold', style=textstyle,\n",
    "                color=color, fontsize=6.5)\n",
    "    \n",
    "def _node_plotter(clade, ax):\n",
    "    confidence = clade.confidence\n",
    "    if confidence is None:\n",
    "        return\n",
    "\n",
    "    if confidence > 0.95:\n",
    "        if clade.is_terminal():\n",
    "            hpos = clade.right - 0.013\n",
    "        else:\n",
    "            hpos = clade.right\n",
    "        color = 'k'\n",
    "        ax.scatter([hpos], [clade.y], color=color,\n",
    "                   marker='o', s=12, linewidths=0, edgecolors='k', zorder=10)\n",
    "\n",
    "    \n",
    "    \n",
    "tree = Phylo.read('data/core.a.mags.muri2.g.final.cds.TIGR02013-hits.sqz.gb.prot.nwk', 'newick')\n",
    "# Rename taxa to species rather than gene.\n",
    "for t in tree.get_terminals():\n",
    "    t.name = t.name.rsplit('_', 1)[0]\n",
    "\n",
    "outgroup = [\n",
    "#               'Porphyromonas_gingivalis_ATCC_33277'\n",
    "            'Bacteroides_ovatus_ATCC_8483'\n",
    "            , 'Bacteroides_thetaiotaomicron_VPI5482'\n",
    "#            , 'Barnesiella_viscericola_DSM_18177'\n",
    "           ]\n",
    "view_outgroup = [\n",
    "#               'Porphyromonas_gingivalis_ATCC_33277'\n",
    "            'Bacteroides_ovatus_ATCC_8483'\n",
    "            , 'Bacteroides_thetaiotaomicron_VPI5482'\n",
    "#            , 'Barnesiella_viscericola_DSM_18177'\n",
    "           ]\n",
    "tree.root_with_outgroup(outgroup)\n",
    "for t in outgroup:\n",
    "    if t not in view_outgroup:\n",
    "        tree.prune(t)\n",
    "\n",
    "tree.clade.branch_length = 0.06\n",
    "#tree.clade.confidence = None\n",
    "tree.collapse_all(lambda c: c.confidence is not None and c.confidence < 0.7)\n",
    "#tree.ladderize(reverse=True)\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(halfwidth, 6))\n",
    "tree_annotated = plot_tree(tree, ax,\n",
    "                           tax_plotter=_tax_plotter,\n",
    "                           node_plotter=_node_plotter)\n",
    "\n",
    "#for confidence_label, color in [('> 95%', 'black'),\n",
    "#                                ('> 90%', 'grey'),\n",
    "#                                ('> 80%', 'lightgrey'),\n",
    "#                                ('< 80%', 'white')\n",
    "#                                ('> 80%', 'white')\n",
    "#                               ]\n",
    "#    ax.scatter([], [], color=color, label=confidence_label, s=50, linewidths=2, edgecolors='k')\n",
    "ax.set_yticks([])\n",
    "ax.set_xticks([])\n",
    "#ax.legend(loc='upper left', title='confidence')\n",
    "\n",
    "scale_xy = (0, 5.5)\n",
    "scale_length = 0.05\n",
    "_x, _y = scale_xy\n",
    "ax.hlines([_y], [_x], [_x + scale_length], lw=1)\n",
    "ax.annotate('{:.1}'.format(scale_length),\n",
    "            xy=(_x + scale_length / 2, _y + 0.4),\n",
    "            ha='center', fontsize=8)\n",
    "\n",
    "ax.axis('off')\n",
    "#fig.tight_layout()\n",
    "\n",
    "plt.savefig('fig/phylogenetics_rpoB.pdf', dpi=250, bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "rename_taxon = {\n",
    "#                  'Barnesiella_viscericola_DSM_18177': 'Bv'\n",
    "#                , 'Bacteroides_ovatus_ATCC_8483': 'Bo'\n",
    "#                , 'Bacteroides_thetaiotaomicron_VPI5482': 'Bt'\n",
    "#                , 'Porphyromonas_gingivalis_ATCC_33277': 'Pg'\n",
    "#                , 'Homeothermus_arabinoxylanisolvens': 'Ha'\n",
    "#                , 'Muribaculum_intestinale_DSM_28989T': 'Mi'\n",
    "#                , 'Duncaniella_muris_DSM_103720T': 'Dm'\n",
    "#                , 'Duncaniella_freteri_DSM_108168T': 'Df'\n",
    "#                , 'Duncaniella_dubosii_DSM_107170T': 'Dd'\n",
    "#                , 'Paramuribaculum_intestinale_DSM_100749T': 'Pi'\n",
    "#                , 'Amulumruptor_caecigallinarius': 'Ac'\n",
    "#                , 'Barnesiella_intestinihominis_YIT_11860': 'Bi'\n",
    "               } \n",
    "                 \n",
    "color_map = { 'starch': 'blue'\n",
    "            , 'host': 'purple'\n",
    "            , 'plant': 'green'\n",
    "            }\n",
    "\n",
    "italic_list = [ 'Barnesiella_viscericola_DSM_18177'\n",
    "              , 'Bacteroides_ovatus_ATCC_8483'\n",
    "              , 'Bacteroides_thetaiotaomicron_VPI5482'\n",
    "              , 'Porphyromonas_gingivalis_ATCC_33277'\n",
    "              , 'Homeothermus_arabinoxylanisolvens'\n",
    "              , 'Muribaculum_intestinale_DSM_28989T'\n",
    "              , 'Duncaniella_muris_DSM_103720T'\n",
    "              , 'Duncaniella_freteri_DSM_108168T'\n",
    "              , 'Duncaniella_dubosii_DSM_107170T'\n",
    "              , 'Paramuribaculum_intestinale_DSM_100749T'\n",
    "              , 'Amulumruptor_caecigallinarius'\n",
    "              , 'Barnesiella_intestinihominis_YIT_11860'\n",
    "              ]\n",
    "\n",
    "con = sqlite3.connect('data/core.muri2.2.denorm.db')\n",
    "\n",
    "mag = (\n",
    "    pd.read_sql(\n",
    "    \"\"\"\n",
    "SELECT *\n",
    "FROM genome\n",
    "JOIN checkm USING (genome_id)\n",
    "JOIN genome_group USING (genome_id)\n",
    "WHERE genome_group_id = 'muri2'\n",
    "    \"\"\",\n",
    "    index_col='genome_id',\n",
    "    con=con,\n",
    "    )\n",
    ")\n",
    "\n",
    "colormap = color_map = { 'ormerod-starch': 'blue'\n",
    "                       , 'ormerod-host': 'purple'\n",
    "                       , 'ormerod-plant': 'green'\n",
    "                       , 'public': 'grey'\n",
    "                       , 'here': 'black'\n",
    "                       , 'outgroup': 'lightgrey'\n",
    "                       }\n",
    "    \n",
    "def _node_plotter(clade, ax):\n",
    "    confidence = clade.confidence\n",
    "    if confidence is None:\n",
    "        return\n",
    "\n",
    "    if confidence > 0.95:\n",
    "        if clade.is_terminal():\n",
    "            hpos = clade.right - 0.013\n",
    "        else:\n",
    "            hpos = clade.right\n",
    "        color = 'k'\n",
    "        ax.scatter([hpos], [clade.y], color=color,\n",
    "                   marker='o', s=12, linewidths=0, edgecolors='k', zorder=10)\n",
    "\n",
    "    \n",
    "    \n",
    "tree = Phylo.read('data/core.a.mags.muri2.g.final.marker_genes.gb.prot.nwk', 'newick')\n",
    "outgroup = [\n",
    "              'Porphyromonas_gingivalis_ATCC_33277'\n",
    "            , 'Bacteroides_ovatus_ATCC_8483'\n",
    "            , 'Bacteroides_thetaiotaomicron_VPI5482'\n",
    "           , 'Barnesiella_viscericola_DSM_18177'\n",
    "           , 'Barnesiella_intestinihominis_YIT_11860'\n",
    "           ]\n",
    "view_outgroup = [\n",
    "           ]\n",
    "\n",
    "tree.root_with_outgroup(outgroup)\n",
    "for t in outgroup:\n",
    "    if t not in view_outgroup:\n",
    "        tree.prune(t)\n",
    "        \n",
    "tree.clade.branch_length = 0.06\n",
    "#tree.clade.confidence = None\n",
    "\n",
    "all_taxa = [t.name for t in tree.get_terminals()]\n",
    "dmat = np.zeros((len(all_taxa), len(all_taxa)))\n",
    "for (i, a), (j, b) in product(list(enumerate(all_taxa)), repeat=2):\n",
    "    if i > j:\n",
    "        continue\n",
    "    else:\n",
    "        d = tree.distance(a, b)\n",
    "        dmat[i, j] = d\n",
    "        dmat[j, i] = d\n",
    "        \n",
    "fit = MDS(n_components=1, metric=True, n_init=10, dissimilarity='precomputed').fit(dmat)\n",
    "embed = pd.DataFrame(fit.embedding_, index=all_taxa).squeeze()\n",
    "cmap = (embed - embed.min()) / (embed.max() - embed.min())\n",
    "cmap = dict(zip(cmap.index, mpl.cm.jet(cmap)))\n",
    "\n",
    "def _tax_plotter(clade, ax):\n",
    "    if clade.name in rename_taxon:\n",
    "        # Check if there's a renaming for a given clade.\n",
    "        label = rename_taxon[clade.name]\n",
    "    else:\n",
    "        label = clade.name\n",
    "        \n",
    "    if mag.subgroup[clade.name] == 'here':\n",
    "        label = label + '*'\n",
    "        \n",
    "    if clade.name in italic_list:\n",
    "        textstyle = 'italic'\n",
    "    else:\n",
    "        textstyle = 'normal'\n",
    "    \n",
    "    color = cmap[clade.name]\n",
    "    ax.annotate(label, xy=(clade.right + 0.015, clade.y),\n",
    "                ha='left', va='center',\n",
    "                weight='bold', style=textstyle,\n",
    "                color=color, fontsize=6.5)\n",
    "\n",
    "tree.collapse_all(lambda c: c.confidence is not None and c.confidence < 0.7)\n",
    "#tree.ladderize(reverse=True)\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(halfwidth, 15))\n",
    "tree_annotated = plot_tree(tree, ax,\n",
    "                           tax_plotter=_tax_plotter,\n",
    "                           node_plotter=_node_plotter)\n",
    "\n",
    "#for confidence_label, color in [('> 95%', 'black'),\n",
    "#                                ('> 90%', 'grey'),\n",
    "#                                ('> 80%', 'lightgrey'),\n",
    "#                                ('< 80%', 'white')\n",
    "#                                ('> 80%', 'white')\n",
    "#                               ]\n",
    "#    ax.scatter([], [], color=color, label=confidence_label, s=50, linewidths=2, edgecolors='k')\n",
    "ax.set_yticks([])\n",
    "ax.set_xticks([])\n",
    "#ax.legend(loc='upper left', title='confidence')\n",
    "\n",
    "scale_xy = (0, 5.5)\n",
    "scale_length = 0.2\n",
    "_x, _y = scale_xy\n",
    "ax.hlines([_y], [_x], [_x + scale_length], lw=1)\n",
    "ax.annotate('{:.1}'.format(scale_length),\n",
    "            xy=(_x + scale_length / 2, _y + 0.4),\n",
    "            ha='center', fontsize=8)\n",
    "\n",
    "panel = 'A'\n",
    "ax.annotate(panel, xy=(0.02, 1.02), xycoords='axes fraction', fontsize=12)\n",
    "\n",
    "\n",
    "ax.axis('off')\n",
    "#fig.tight_layout()\n",
    " \n",
    "#plt.savefig('fig/phylogenetics_concatenated.pdf', dpi=250, bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# TODO: Make a table indicating the presence/abscence of a long list of OPFs w/(out) particular localization\n",
    "# Use this table to place dots right of the tip labels (potentially with dotted lines indicating how stuff lines up.)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}